src/libraries/System.Private.CoreLib/src/System/IO/WindowsFileStreamStrategy.cs

```diff
@@ -77,7 +82,25 @@ internal WindowsFileStreamStrategy(string path, FileMode mode, FileAccess access
 
         public sealed override bool CanWrite => !_fileHandle.IsClosed && (_access & FileAccess.Write) != 0;
 
-        public unsafe sealed override long Length => FileStreamHelpers.GetFileLength(_fileHandle, _path);
+        public unsafe sealed override long Length => _share > FileShare.Read ?
+            FileStreamHelpers.GetFileLength(_fileHandle, _path) :
+            _length ??= FileStreamHelpers.GetFileLength(_fileHandle, _path);
+
+        protected void UpdateLengthOnChangePosition()
+        {
+            // Do not update the cached length if the file can be written somewhere else
+            // or if the length has not been queried.
+            if (_share > FileShare.Read || _length is null)
+            {
+                Debug.Assert(_length is null);
```

Comment:
> _or if the length has not been queried_

If I'm interpreting this correctly, the above statement is saying that `_length is null` is a valid state. If that's the case, is this assert necessary? Or what am I missing?

Reply:
Is meant to verify that `_length` is never populated/used when the file share (`_share` ) does not allow us to cache the file length.

Reply:
Oh, right! If `_share` is either `FileShare.None` or `FileShare.Read`, we expect `_length` to never be `null` at this point.

Reply:
Yes but the purpose of the assert is to check that _buffer is always `null` for the case where `_share` is other than `None` or `Read`. Since the OR operator is going to short-circuit with the first expression I wanted to verify the second expression anyways.